src/macros.jl

```diff
@@ -278,6 +265,49 @@ function assert_validmodel(m, macrocode)
     end
 end
 
+"""
+    macro_return(model, code, variable)
+
+Return a code that runs `code` in a local scope which returns the value of `variable`.
+"""
+function macro_return(code, variable)
+    return quote
+        let
+            # The let block ensures that all variables created behaves like
+            # local variables, see https://github.com/JuliaOpt/JuMP.jl/issues/1496
+            # To make $variable accessible from outside we need to return it at
+            # the end of the block
+            $code
+            $variable
+        end
+    end
+end
+
+"""
+    macro_assign_and_return(model, code, variable, registerfun::Function,
+                            quotvarname, escvarname, final_variable=variable)
+
+Return runs `code` in a local scope which returns the value of `variable`
+and then assign `final_variable` to `escvarname`.
```

Comment:
Could we take the `varname` and `quotvarname` as a single symbol and escape/quote it as needed?

Reply:
Indeed, I was able to get rid of `quotvarname` and `escvarname`. It makes things more readable to isolate this technical details in macro_assign_and_return :)